### PR TITLE
MGMT-16199: block unbind host while cluster is progressing

### PR DESCRIPTION
**Pull request by eifrach**
bugfix
-  remove the ability to unbind Host while the cluster is progressing 
- allow removing host's which are not bound to a cluster , for ACM 

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/common/common_test.go [hunk 1]
```diff
@@ -466,6 +466,55 @@ var _ = Describe("JSON serialization checks", func() {
...
+				models.ClusterStatusFinalizing,
+				models.ClusterStatusInstalling,
+				models.ClusterStatusInstallingPendingUserAction,
+				models.ClusterStatusPreparingForInstallation,
+			}
+			for _, failState := range notAllowedStatus {
+				cluster.Status = swag.String(failState)
+				err := CanUnbindhost(&cluster)
+
+				Expect(err).ToNot(BeNil())
```
Comment (paul-maidment) [inline internal/common/common_test.go:496]: Whereas this checks that there is error handling generally it doesn't verify that the error here https://github.com/openshift/assisted-service/pull/5809/files#diff-e0a57b903e60761f673a4485b4a747f500f4155828e182cc19786ee5648cd6e6R448 was the one emitted. 

I'd personally update the test to verify that the expected error is returned.

File: internal/common/common.go [hunk 2]
```diff
@@ -432,6 +435,22 @@ func GetInventoryInterfaces(inventory string) (string, error) {
 	return inventory[interfacesLocation : interfacesLocation+endLocation], nil
 }
 
+// blocking host removal while cluster is progressing
+func CanUnbindhost(c *Cluster) (err error) {
+	clusterStatus := swag.StringValue(c.Status)
+	NotAllowedStatuses := []string{
+		models.ClusterStatusFinalizing,
+		models.ClusterStatusInstalling,
+		models.ClusterStatusInstallingPendingUserAction,
```
Comment (tsorya) [inline internal/common/common.go:444]: I can think about scenario were i would actually like to remove node in ClusterStatusInstallingPendingUserAction.
Lets say worker is stuck in it and i have more workers wouldn't i want to continue installation and not handle stuck worker? @avishayt ?

Comment (paul-maidment) [reply internal/common/common.go:444]: I agree, this seems like a likely use case for wanting to remove the node.

File: internal/host/host.go [hunk 3]
```diff
@@ -1441,15 +1447,23 @@ func (m *Manager) GetHostByKubeKey(key types.NamespacedName) (*common.Host, erro
 }
 
 func (m *Manager) UnRegisterHost(ctx context.Context, h *models.Host) error {
-	if err := common.DeleteHostFromDB(m.db, h.ID.String(), h.InfraEnvID.String()); err != nil {
-		return err
-	}
 	if h.ClusterID != nil {
```
Comment (tsorya) [inline internal/host/host.go:1450]:
```suggestion
	if h.ClusterID == nil {
	     return err
	 }
	
```

File: internal/common/common.go (hunk 2 above)
Comment (avishayt) [reply internal/common/common.go:444]: It leave some room for people to shoot themselves in the foot, but I'd be OK with that compromise.

File: internal/common/common_test.go (hunk 1 above)
Comment (eifrach) [reply internal/common/common_test.go:496]: correct, it does verify the exact error, just check if we go err back.
I'll added also the error string check 
thanks

File: internal/common/common.go (hunk 2 above)
Comment (eifrach) [reply internal/common/common.go:444]: I think we should advise the user to cancel the installation and then remove nodes.

so when, in `ClusterStatusInstallingPendingUserAction` we are still progressing the installation thus we don't want user to remove nodes.

File: internal/host/host.go (hunk 3 above)
Comment (eifrach) [reply internal/host/host.go:1450]: I do return an error at the end.  do you think it better this way ?

File: internal/host/host.go [hunk 4]
```diff
@@ -1441,14 +1447,29 @@ func (m *Manager) GetHostByKubeKey(key types.NamespacedName) (*common.Host, erro
...
+			m.log.Infof("Host %v is not Bound to Cluster.", h.ID)
+		}
 	}
 	if h.ClusterID != nil {
-		if err := m.db.Model(&common.Cluster{}).Where("id = ?", h.ClusterID).Update("trigger_monitor_timestamp", time.Now()).Error; err != nil {
+		var cluster *common.Cluster
+		if err := m.db.Select("status").Take(&cluster, "id = ?", h.ClusterID.String()).Error; err != nil {
+			return fmt.Errorf("failed to get cluster from DB: %+v", err)
+		}
+		if err := common.CanUnbindhost(cluster); err != nil {
```
Comment (carbonin) [inline internal/host/host.go:1462]: This function `UnRegisterHost` is for deleting hosts, not unbinding them.

If you want this to be run during unbind you should call it from `UnbindHost` if it's checking if a host can be deleted it should have a different name.

**Review by carbonin — CHANGES_REQUESTED** [review]

**Comment by carbonin** [comment]
Also it would be much easier to review if you stick to only making changes that are related to your bug fix.

All the formatting changes are unnecessary in this PR.

**Pull request closed without merging**